File: src/main/java/in/costea/wiles/data/CompilationExceptionsCollection.java
package in.costea.wiles.data;

import in.costea.wiles.exceptions.CompilationException;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import static in.costea.wiles.statics.Constants.DEBUG;

public class CompilationExceptionsCollection extends ArrayList<CompilationException>
{

    public CompilationExceptionsCollection(CompilationExceptionsCollection exceptions)
    {
        super(exceptions);
    }

    public CompilationExceptionsCollection()
    {
        super();
    }

    public void add(@NotNull CompilationException... exceptions)
    {
        this.addAll(List.of(exceptions));
    }

    public void add(@NotNull CompilationExceptionsCollection objToAdd)
    {
        this.addAll(objToAdd);
    }

    public String getExceptionsString()
    {
        var optional = stream().sorted((a, b) ->
                {
                    var loc1 = a.getLocation();
                    var loc2 = b.getLocation();
                    if (loc1 == null) return 1;
                    if (loc2 == null) return -1;
                    var comp = Integer.compare(loc1.line(), loc2.line());
                    if (comp != 0) return comp;
                    var comp2 = Integer.compare(loc1.lineIndex(), loc2.lineIndex());
                    if (comp2 != 0) return comp2;
                    return a.getMessage().compareTo(b.getMessage());
                }).
                map((Exception x) -> "\n    " + x.getMessage() + "\n" + (DEBUG ? Arrays.toString(x.getStackTrace()) : "")).
                reduce((a, b) -> a + b);
        if (optional.isEmpty())
            throw new IllegalStateException();
        return optional.get();
    }

    @Override
    public boolean equals(Object o)
    {
        if (o instanceof CompilationExceptionsCollection obj)
        {
            if (size() != obj.size())
                return false;
            for (int i = 0; i < size(); i++)
                if (!Objects.equals(get(i).getMessage(), obj.get(i).getMessage()))
                    return false;
            return true;
        }
        return false;
    }
}

File: src/main/java/in/costea/wiles/data/CompilationExceptionsCollection.kt
package `in`.costea.wiles.data

import `in`.costea.wiles.constants.ErrorMessages.COMPILATION_FAILED_ERROR
import `in`.costea.wiles.constants.ErrorMessages.LINE_SYMBOL
import `in`.costea.wiles.constants.Settings.DEBUG
import `in`.costea.wiles.exceptions.AbstractCompilationException
import `in`.costea.wiles.exceptions.InternalErrorException

class CompilationExceptionsCollection : ArrayList<AbstractCompilationException>() {
    fun getExceptionsString(input: String): String {
        val optional = sortedWith(nullsLast(compareBy<AbstractCompilationException> { it.tokenLocation.line }
            .thenBy { it.tokenLocation.lineIndex }))
            .map { LINE_SYMBOL+ "Line ${it.tokenLocation.line}: " + it.message +
                    it.tokenLocation.displayLocation(input) + (if (DEBUG) "\n"+it.stackTraceToString() else "") }
            .fold("") { a, b -> a + b }
        if (optional.isEmpty())
            throw InternalErrorException()
        return COMPILATION_FAILED_ERROR+optional
    }

    override fun equals(other: Any?): Boolean {
        if (other !is CompilationExceptionsCollection)
            return false
        if (other.size != size)
            return false
        for (i in 0 until size)
            if (this[i].message != other[i].message)
                return false
        return true
    }

    override fun hashCode(): Int {
        var hashCode = 7
        for (obj in this)
            hashCode = 31 * hashCode + obj.message.hashCode()
        return hashCode
    }

    fun throwFirstIfExists()
    {
        if(size>0)
            throw this[0]
    }

}
